File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun removeNthFromEnd(head: ListNode?, n: Int): ListNode? {
        val result = ListNode(0)
        result.next = head

        var ptr1: ListNode? = result
        var ptr2: ListNode? = head

        for(i in 0 until n) {
            ptr2 = ptr2?.next
        }

        while(ptr2 != null) {
            ptr2 = ptr2?.next
            ptr1 = ptr1?.next
        }

        ptr1?.next = ptr1?.next?.next

        return result?.next
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        first = head
        second = slow.next

        slow.next = None
        
        second = self.reverse(second)

        while first and second:
            tmp1, tmp2 = first.next, second.next

            first.next = second
            second.next = tmp1

            first = tmp1
            second = tmp2
        
        
    def reverse(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current = head
        prev = None

        while current:
            tmp = current.next
            current.next = prev
            prev = current
            current = tmp

        return prev